### ISSUE TITLE
Feature-request: IO-limiting mode for `get` operations

### ISSUE THREAD

**Issue by mih** [datalad/datalad#3521]
## Scenario

There is a dataset with a few (<30) files that are each huge (each is several dozen GB of https://www.awaresystems.be/imaging/tiff/bigtiff.html). The "normal" procedure to capture the (comparatively) small results of an analysis of these data is to

- create a new dataset for the results
- install the bigtiff dataset as a subdataset
- use `datalad run` with appropriate `--input` args to execute the analysis

## Problem

This procedure creates a substantial IO bottleneck. Even if all IO is happening locally on a machine (shared file system), about 1 *TB* of data needs to be copied -- which takes time, a sizable chunk of additional storage, and causes noticeable churn on the filesystem. In this scenario, this copy has a lifetime of only a few minutes (the run time of the analysis). Afterwards, it will be thrown away.

## Potential solutions

We already have `install --reckless` that was motivated by a similar desire (fast read-only access, previous discussion https://github.com/datalad/datalad/issues/1):

```
  --reckless      Set up the dataset to be able to obtain content in the
                        cheapest/fastest possible way, even if this poses a
                        potential risk to data integrity (e.g. hardlink files
                        from a local clone of the dataset). Use with care, and
                        limit to "read-only" use cases. With this flag the
                        installed dataset will be marked as untrusted.
                        [Default: False]
```

which ultimately sets `annex.hardlink=true` and `untrust`s that clone. This can help if the data is on the same filesystem, but is helpless if not. In the concrete case that motivates this issue, we are on HPC infrastructure, and the input data lives on a dedicated filesystem (GPFS) that is providing relatively static data to an army of compute nodes.

What about populating `.git/annex/objects` of the temporary input dataset clone with **symlinks** to the keys in `.git/annex/objects` of the source dataset on the other file system? This would constrain storage impact to the number of inodes, but have little to no cost re size and bandwidth.
Or even have `.git/annex/objects` itself be a straight-up symlink to the entire annex tree of the source dataset? In that case, we could even save a (possibly substantial) number of required inodes on the FS (not much of an issue in this concrete scenario, but a general concern with many HPC setups). Of course, this is only meaningful for read-only operations (and writes could be prevented by appropriate permission constraints on the source dataset directories).

@bpoldrack will shortly report on an exploration on how git-annex reacts to such a setup. 

@joeyh it would be great to get your thoughts on this approach for a cheap, temporary, read-only access.

**Comment by bpoldrack** [comment]
That approach seems to work just fine with symlinking from within the target object tree into the source object tree as well as just symlinking the entire `.git/annex/object`. All it takes to make annex aware is a `git annex fsck` afterwards.

<details>
<summary>Exploration for symlinking the entire object tree:</summary>

First, set up an original data source to provide some keys:
```
/tmp$ datalad create origin
[INFO   ] Creating a new annex repo at /tmp/origin 
create(ok): /tmp/origin (dataset)
/tmp$ cd origin
/tmp/origin$ echo "some" > file_a
/tmp/origin$ echo "other" > file_b
/tmp/origin$ datalad save -m "provide some content"
add(ok): file_a (file)                                                                                                       
add(ok): file_b (file)
save(ok): . (dataset)
/tmp/origin$ cd ..
```

Now, have the actual source dataset (target of the symlink we want to build later on), that'll have a file present and another not:
```
/tmp$ datalad install -s origin source
[INFO   ] Cloning origin into '/tmp/source' 
install(ok): /tmp/source (dataset)
/tmp$ cd source
/tmp/source$ datalad get file_a
get(ok): /tmp/source/file_a (file) [from origin...]
/tmp/source$ cd ..
```
Finally build our test dataset: 
```
/tmp$ datalad install -s source test
[INFO   ] Cloning source into '/tmp/test' 
install(ok): /tmp/test (dataset)
/tmp$ cd test
/tmp/test$ ln -s /tmp/source/.git/annex/objects .git/annex/objects
/tmp/test$ tree .git/annex
.git/annex
├── index
├── index.lck
├── journal
├── journal.lck
├── mergedrefs
├── objects -> /tmp/source/.git/annex/objects
├── othertmp.lck
├── sentinal
└── sentinal.cache

2 directories, 7 files
/tmp/test$ git annex whereis file_a
whereis file_a (2 copies) 
        b66ae28f-6d1f-4821-a20a-da5c739ace20 -- ben@tree:/tmp/source [origin]
        dc8f097e-e918-4567-9b62-0407b1ba6463 -- ben@tree:/tmp/origin
ok
/tmp/test$ git annex whereis file_b
whereis file_b (1 copy) 
        dc8f097e-e918-4567-9b62-0407b1ba6463 -- ben@tree:/tmp/origin
ok
```
And make annex aware:
```
/tmp/test$ git annex fsck
fsck file_a (fixing location log) (checksum...) ok
fsck file_b ok
(recording state in git...)
/tmp/test$ tree .git/annex
.git/annex
├── index
├── index.lck
├── journal
├── journal.lck
├── mergedrefs
├── objects -> /tmp/source/.git/annex/objects
├── othertmp.lck
├── sentinal
└── sentinal.cache

2 directories, 7 files
/tmp/test$ git annex whereis file_a
whereis file_a (3 copies) 
        96b10e95-378f-4ab3-9932-86d9c265268d -- ben@tree:/tmp/test [here]
        b66ae28f-6d1f-4821-a20a-da5c739ace20 -- ben@tree:/tmp/source [origin]
        dc8f097e-e918-4567-9b62-0407b1ba6463 -- ben@tree:/tmp/origin
ok
/tmp/test$ git annex whereis file_b
whereis file_b (1 copy) 
        dc8f097e-e918-4567-9b62-0407b1ba6463 -- ben@tree:/tmp/origin
ok
```
</details>

**Comment by bpoldrack** [comment]
FTR: Using `git annex fsck --fast` instead seems to be the better approach, of course, and that works just as well in the examples above.

**Comment by joeyh** [comment]
Symlinking .git/annex will work. Symlinking only .git/annex/objects will cause
git-annex to be unusually slow in some situations when it is unable to
rename() from .git/annex to there.

Symlinking individual object files has been proposed before at some
point. IIRC my problem with it was it seemed likely to complicate a
large number of edge cases.

-- 
see shy jo

**Comment by joeyh** [comment]
Might a git-worktree(1) rather than a clone be a better way to approach
this?

-- 
see shy jo

**Comment by mih** [comment]
@joeyh Thanks for the feedback!

A worktree would be the optimal choice IMHO. However, IIRC GitPython had some issues with that -- although that might be outdated knowledge by now. So until we have managed to complete a relatively substantial RF to get it out of datalad, we are somewhat limited in what we can do (https://github.com/datalad/datalad/issues/2879).

Given that we have identified one immediately workable path, and a potential future solution, I think we can close this issue. I have created https://github.com/datalad/datalad/issues/3528 with a more focused scope.
